processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/MostSpecificResultTypeSelector.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.internal.model.source.selector;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.mapstruct.ap.internal.model.common.Type;
+import org.mapstruct.ap.internal.model.source.Method;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class MostSpecificResultTypeSelector implements MethodSelector {
+
+    @Override
+    public <T extends Method> List<SelectedMethod<T>> getMatchingMethods(Method mappingMethod,
+                                                                         List<SelectedMethod<T>> candidates,
+                                                                         List<Type> sourceTypes, Type mappingTargetType,
+                                                                         Type returnType, SelectionCriteria criteria) {
+        if ( candidates.size() < 2 || !criteria.isForMapping() || criteria.getQualifyingResultType() != null) {
+            return candidates;
+        }
+
+        List<SelectedMethod<T>> result = new ArrayList<>();
+
+        for ( SelectedMethod<T> candidate : candidates ) {
+            if ( candidate.getMethod()
+                .getResultType()
```

Comment:
since you specifically use the resultType.. should we have an update method test?

Reply:
Good point. I'll try to add such test case